### PR TITLE
[GUILD-3142] - Allowlist file processing

### PR DESCRIPTION
**Pull request by cs-balazs**
[Linear](https://linear.app/guildxyz/issue/GUILD-3142/fe-implementation)

- Upload allowlist files to GCS
- The uploaded file's ID is saved in the requirement's `.data.fileId` field
- The BE sends information on the processing status, and the number of processed addresses

> merge only once gate and core are both merged, and they work as expected

### CODE REVIEW

File: src/requirements/Allowlist/components/AllowlistFormInputs.tsx [hunk 1]
```diff
@@ -48,103 +53,194 @@ export default function AllowlistFormInputs({
...
+                uploadFileToGcs.reset()
+                setValue(`${baseFieldPath}.data.addresses`, [], {
+                  shouldDirty: true,
+                })
+                setValue(`${baseFieldPath}.data.fileId`, undefined, {
+                  shouldDirty: true,
+                })
+              }}
+            >
+              clear
```
Comment (BrickheadJohnny) [inline src/requirements/Allowlist/components/AllowlistFormInputs.tsx:143]: Let's use an uppercase `C` here & I think we could also add an icon ;)

File: src/requirements/Allowlist/components/AllowlistFormInputs.tsx [hunk 2]
```diff
@@ -48,103 +53,194 @@ export default function AllowlistFormInputs({
...
 
   return (
     <Stack spacing={4} alignItems="start" {...getRootProps()}>
       <FormControl isInvalid={!!fileRejections?.[0]} textAlign="left">
-        <FormLabel>Upload from file</FormLabel>
+        <FormLabel display={"flex"} gap={2}>
+          Upload from file
+          {addressCount && (
+            <Tooltip
+              label={`Currently an address with ${addressCount} addresses is uploaded. Uploading a new file overwrites these addresses`}
```
Comment (BrickheadJohnny) [inline src/requirements/Allowlist/components/AllowlistFormInputs.tsx:124]: I think this should be "Currently an allowlist...", hehe? 👀

File: src/requirements/Allowlist/components/AllowlistFormInputs.tsx [hunk 3]
```diff
@@ -48,103 +53,194 @@ export default function AllowlistFormInputs({
...
+            >
+              <Tag>{addressCount}</Tag>
```
Comment (BrickheadJohnny) [inline src/requirements/Allowlist/components/AllowlistFormInputs.tsx:126]: Small change, but maybe we could use `Intl.NumberFormat` to format the number so it is more readable, what do you think?

File: src/requirements/Allowlist/AllowlistRequirement.tsx [hunk 4]
```diff
@@ -75,6 +90,23 @@ const AllowlistRequirement = ({ ...rest }: RequirementProps): JSX.Element => {
         </Button>
       )}
       <SearchableVirtualListModal
+        aboveList={
```
Comment (BrickheadJohnny) [inline src/requirements/Allowlist/AllowlistRequirement.tsx:93]: I was about to say that maybe we could maybe do this in a more composable way, but that would require a `SearchableVirtualListModal` refactor, and afaik this is not the final solution for this (you mentioned some kind of checkmark icon/animation if the address is included in the allowlist), so we can keep it this way for now.

File: src/requirements/Allowlist/components/AllowlistFormInputs.tsx [hunk 5]
```diff
@@ -48,103 +53,194 @@ export default function AllowlistFormInputs({
...
+              validate:
+                type === "ALLOWLIST_EMAIL"
+                  ? validateEmailAllowlist
+                  : (value_) => {
+                      if (!value_) return
+
+                      const validAddresses = value_.filter(
+                        (address) => address !== "" && isValidAddress(address)
+                      )
+                      // for useBalancy
```
Comment (BrickheadJohnny) [inline src/requirements/Allowlist/components/AllowlistFormInputs.tsx:203]: haha

File: src/requirements/Allowlist/components/AllowlistFormInputs.tsx (hunk 1 above)
Comment (cs-balazs) [reply src/requirements/Allowlist/components/AllowlistFormInputs.tsx]: Good idea! Made some changes in https://github.com/guildxyz/guild.xyz/pull/1244/commits/f69a44165e6801f02af2afd7fc797f7f99002c9a!

File: src/requirements/Allowlist/components/AllowlistFormInputs.tsx (hunk 3 above)
Comment (cs-balazs) [reply src/requirements/Allowlist/components/AllowlistFormInputs.tsx]: Agreed, changed in https://github.com/guildxyz/guild.xyz/pull/1244/commits/7c653fecfde6144bf16f711a9262998f72501104!

File: src/requirements/Allowlist/components/AllowlistFormInputs.tsx (hunk 2 above)
Comment (cs-balazs) [reply src/requirements/Allowlist/components/AllowlistFormInputs.tsx]: Nice catch 😄 
https://github.com/guildxyz/guild.xyz/pull/1244/commits/1e66494587c36314e664d0c9d4246f1ac0e2af9b

File: src/requirements/Allowlist/AllowlistRequirement.tsx (hunk 4 above)
Comment (cs-balazs) [reply src/requirements/Allowlist/AllowlistRequirement.tsx:93]: Good idea, thanks for sharing! We should brainstorm on this, once we have a clear idea on what to do with the modal!

File: src/requirements/Allowlist/components/AllowlistFormInputs.tsx (hunk 1 above)
Comment (BrickheadJohnny) [reply src/requirements/Allowlist/components/AllowlistFormInputs.tsx]: Nice! Btw I think the `Button` component has an `iconSpacing` prop ;)

File: src/requirements/Allowlist/AllowlistRequirement.tsx [hunk 6]
```diff
@@ -32,6 +33,9 @@ const AllowlistRequirement = ({ ...rest }: RequirementProps): JSX.Element => {
 
   const { addresses: initialAddresses, hideAllowlist } = requirement.data
 
+  // These are not included in the schemas, as these are appended on-the-fly by the BE, when sending the response
```
[the remark this answers is not in the document]
Comment (cs-balazs) [reply src/requirements/Allowlist/AllowlistRequirement.tsx]: Fantastic idea! Changed in https://github.com/guildxyz/guild.xyz/pull/1244/commits/232c922d53dfab4c4b05bcc5f3f64687bd506c53!

File: src/components/create-guild/Requirements/components/ExistingRequirementEditableCard.tsx [hunk 7]
```diff
@@ -123,6 +123,15 @@ const ExistingRequirementEditModal = ({
   const { onSubmit: onEditRequirementSubmit, isLoading: isEditRequirementLoading } =
     useEditRequirement(requirement.roleId, {
       onSuccess: (editedRequirement) => {
+        if (
```
[the remark this answers is not in the document]
Comment (cs-balazs) [reply src/components/create-guild/Requirements/components/ExistingRequirementEditableCard.tsx]: I agree, would be better inside the hook, made the change in https://github.com/guildxyz/guild.xyz/pull/1244/commits/95e0ba6e5ac3b45d2b6ab10995dc6ccf3defe120
I'll test it on dev before resolving :ayy:

File: src/requirements/Allowlist/components/AllowlistFormInputs.tsx (hunk 1 above)
Comment (cs-balazs) [reply src/requirements/Allowlist/components/AllowlistFormInputs.tsx]: Nice, thanks, changed to that prop in https://github.com/guildxyz/guild.xyz/pull/1244/commits/bf7dabecf0b8369366828e9a4f136179b8876bdd!

**Review by BrickheadJohnny — APPROVED** [review]
LFG 🚀

**Pull request merged**